### ISSUE TITLE
Pressing mouse button outside of widget area pushes the button widget.

### ISSUE THREAD

**Issue by beru** [cnjinhao/nana#67]
Hi, I noticed pressing mouse button outside of widget area pushes the button widget.
Below are the steps to reproduce it.

1. Press a button widget with mouse click and do not release it. A bit strangely, it reacts to any (left/middle/right) mouse button.
2. Keep pressing mouse button and move mouse cursor to outside of the button widget.
3. Press another mouse button while pressing the first mouse button.
4. The button widget is pushed while mouse button is pressed. Luckily, click event is not fired.

**Comment by cnjinhao** [comment]
Thanks for reporting this issue, a fix has been checked into the develop branch. It fixed an issue of arg_mouse, added a new member arg_mouse::button to determines a button which triggers the event. The arg_mouse::left/mid/right_button indicate the state of buttons while event is emitted.

**Comment by beru** [comment]
Hi, Jinhao. Thank you for the fix. I confirmed the issues I reported are fixed.

I've noticed other things.

- Mouse middle button click and right button click on button widget switches widget focus to the button. Typical Windows applications do not react to those.

- Pressing space key or enter key fires click event of the focused button widget, but appearence of the button does not change in current develop branch build. So It would be better if picture of the button widget changes as if it is pushed by mouse.

**Comment by cnjinhao** [comment]
Hi, thank you for your feedback. A fix has been checked into the develop branch. This fix brought some breaking changes. The parameter type of click event is changed from arg_mouse to arg_click. Just clicking mouse left button or pressing space key(depend on widget) can emit the click event.

**Comment by beru** [comment]
Hi, I've pulled and checked the change.
I confirmed that picture of button widget changes with space and enter keydown.

Strangely, appearence of multiple buttons can be pushed at the same time with following procedures.
- Push a button with space key and keep pressing the button by not releasing space key.
- Push another button with lett mouse button pushdown.
- Release space key or release left mouse button
- First button's picture is still being pressed.

And it seems to me non-focused button widgets react to middle and right mouse button pressing. Button widget's border illumination effects goes off when mouse button is pressed.

I guess state management of GUI widgets is surely a tough thing. So I predict it would take some time to completely fix these problems.
